src/ssh-generator/ssh-generator.c

```diff
@@ -0,0 +1,476 @@
+                        return 0;
+                }
+                if (errno == EROFS || ERRNO_IS_PRIVILEGE(errno)) {
+                        log_debug("Container manager does not provide write access to /run/host/unix-export/, not binding AF_UNIX socket there.");
+                        return 0;
+                }
+
+                return log_debug_errno(errno, "Unable to check if /run/host/unix-export exists: %m");
+        }
+
+        r = make_sshd_template_unit(
+                        dest,
+                        "sshd-unix-export@.service",
+                        sshd_binary,
+                        found_sshd_template_unit,
+                        generated_sshd_template_unit);
+        if (r < 0)
+                return r;
+
+        r = write_socket_unit(
+                        dest,
+                        "sshd-unix-export.socket",
+                        "/run/host/unix-export/ssh",
+                        "AF_UNIX Export");
+        if (r < 0)
+                return r;
+
+        log_info("Binding SSH to AF_UNIX socket /run/host/unix-export/ssh\n"
+                 "→ connect via 'ssh unix/run/systemd/nspawn/unix-export/\?\?\?/ssh' from host");
+
+        return 0;
+}
+
+static int add_extra_sockets(
+                const char *dest,
+                const char *sshd_binary,
+                const char *found_sshd_template_unit,
+                char **generated_sshd_template_unit) {
+
+        unsigned n = 1;
+        int r;
+
+        assert(dest);
+        assert(sshd_binary);
+        assert(generated_sshd_template_unit);
+
+        if (strv_isempty(arg_listen_extra))
+                return 0;
+
+        STRV_FOREACH(i, arg_listen_extra) {
+                _cleanup_free_ char *service = NULL, *socket = NULL;
+
+                if (n > 1) {
+                        if (asprintf(&service, "sshd-extra-%u@.service", n) < 0)
+                                return log_oom();
+
+                        if (asprintf(&socket, "sshd-extra-%u.socket", n) < 0)
+                                return log_oom();
+                }
+
+                r = make_sshd_template_unit(
+                                dest,
+                                service ?: "sshd-extra@.service",
+                                sshd_binary,
+                                found_sshd_template_unit,
+                                generated_sshd_template_unit);
+                if (r < 0)
+                        return r;
+
+                r = write_socket_unit(
+                                dest,
+                                socket ?: "sshd-extra.socket",
+                                *i,
+                                *i);
+                if (r < 0)
+                        return r;
+
+                log_info("Binding SSH to socket %s.", *i);
+                n++;
+        }
+
+        return 0;
+}
+
+static int parse_credentials(void) {
```

Comment:
This part just seems ugly. I'd much prefer if we parse a glob credential. (I'm also fine with not having this for now if that needs more credential infra to be added)

Reply:
I think we really should consider credentials to be more like files, not like individual config options. Because that's ultimately what they are going to end up on disk. This is systematically different from a kernel command line option, which is is not a file, but a single setting that might appear in a file.

This follows the logic we already implemented with tmpfiles.extra and sysusers.extra i guess: we read one file with multiple entries. Same here: we read one file with multiple entries. And the file can have comments and stuff.

So, I wouldn't go overboard with the glob stuff. Use it where necessary, but by default consider things flat and have stable names for things.

Reply:
It appears to me this is something to deal with in the nspawn/vmspawn side, i.e. add an option to concat things when loading/setting an credential. i.e. let's not conflate two problems:

1. the need to configure different objects of the same type
2. the need to configure the same object but from multiple source files

I think the globs stuff is great for the first issue, but for the 2nd issue we should find another solution.

Or to say this differently: an insular solution for `ssh.listen` sounds wrong to me since the same exists for `sysusers.extra`, `fstab.exta`, `tmpfiles.extra`, `login.issue`, `login.motd`, `getty.ttys.serial`, `getty.ttys.container`, `sysctl.extra`, `ssh.authorized_keys.root`, `network.hosts`, `networkd.dns`, `network.search_domains`, the same.

For example we could say "LoadCredential=+" means: append if already set, or so.

Reply:
and one thing i wanted to underline: all the creds that we already read that could be exended via append use newlines as separators, which is really nice, because unlike separators like spaces, or commas or so files tend to have newlines at the end of the files, hence appending multiple of them simply on the byte level typically yields fully valid results, which it wouldn't if we used other separators.

(i realized this the first time when i added getty.ttys.serial, because i was wondering back then which separator to use.)

We should really keep this in mind: for creds that can be appended, we should focus on using newlines as separators, i.e. list items in separate lines rather than in comma separated lines